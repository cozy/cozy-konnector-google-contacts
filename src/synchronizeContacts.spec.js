const getCozyToGoogleStrategy = require('./getCozyToGoogleStrategy')
const synchronizeContacts = require('./synchronizeContacts')
const googleUtils = require('./google')
const { mockDate, restoreDate } = require('../jestHelpers/mockDate')

jest.mock('./google')

const SOURCE_ACCOUNT_ID = '45c49c15-4b00-48e8-8bfd-29f8177b89ff'
const OTHER_SOURCE_ACCOUNT_ID = 'cb31eb3e-989e-4818-8b45-afed904237da'

const MOCKED_DATE = '2018-05-05T09:09:00.115Z'

const cozyContacts = [
  {
    id: 'jane-doe-attached-to-the-source',
    name: { givenName: 'Jane', familyName: 'Doe' },
    cozyMetadata: {
      sync: {
        [SOURCE_ACCOUNT_ID]: {
          id: 'people/123456'
        }
      }
    }
  },
  {
    id: 'john-doe-attached-to-the-source',
    name: { givenName: 'John', familyName: 'Doe' },
    cozyMetadata: {
      sync: {
        [SOURCE_ACCOUNT_ID]: {
          id: 'people/987654'
        }
      }
    }
  },
  {
    // a contact that is not attached to a source
    id: 'reinhold-jenkins-no-source',
    name: { givenName: 'Reinhold', familyName: 'Jenkins' },
    cozyMetadata: {
      updatedByApps: ['Contacts']
    }
  },
  {
    id: 'larue-cremin-attached-to-another-source',
    name: { givenName: 'Larue', familyName: 'Cremin' },
    cozyMetadata: {
      sync: {
        [OTHER_SOURCE_ACCOUNT_ID]: {
          id: 'people/987654'
        }
      }
    }
  }
]

beforeAll(() => {
  mockDate(MOCKED_DATE)
})

afterAll(() => {
  restoreDate()
})

describe('synchronizeContacts', () => {
  const fakeCozyClient = {}

  beforeEach(() => {
    googleUtils.createContact.mockResolvedValueOnce({
      data: {
        etag: 'etag-963fd240-d568-4b24-9f26-6e658d4e4958',
        resourceName: 'people/96508'
      }
    })
    googleUtils.createContact.mockResolvedValueOnce({
      data: {
        etag: 'etag-6deab8dd-5a0f-451d-bcfe-fd81dd57bdcd',
        resourceName: 'people/78485'
      }
    })
    fakeCozyClient.save = jest.fn(contact =>
      Promise.resolve({
        data: { id: contact.id }
      })
    )
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('synchronize contacts with cozy to google strategy', () => {
    it('should synchronize cozy contacts to google', async () => {
      const googleContacts = [] // we don't care about google contacts here
      const strategy = getCozyToGoogleStrategy(
        fakeCozyClient,
        googleUtils,
        SOURCE_ACCOUNT_ID
      )
      const result = await synchronizeContacts(
        cozyContacts,
        googleContacts,
        strategy
      )
      expect(result).toEqual([
        null,
        null,
        { created: true, id: 'reinhold-jenkins-no-source' },
        { created: true, id: 'larue-cremin-attached-to-another-source' }
      ])

      expect(fakeCozyClient.save).toHaveBeenCalledTimes(2)
      expect(fakeCozyClient.save.mock.calls[0]).toMatchSnapshot()
      expect(fakeCozyClient.save.mock.calls[1]).toMatchSnapshot()
    })

    it('should do nothing if there are no contacts to sync', async () => {
      const cozyContacts = []
      const googleContacts = []
      const strategy = getCozyToGoogleStrategy(
        fakeCozyClient,
        googleUtils,
        SOURCE_ACCOUNT_ID
      )
      const result = await synchronizeContacts(
        cozyContacts,
        googleContacts,
        strategy
      )
      expect(result).toEqual([])
    })

    it('should fail nicely on google error', async () => {
      const googleContacts = []
      googleUtils.createContact.mockRejectedValue(
        'Unable to create google contact'
      )
      const strategy = getCozyToGoogleStrategy(
        fakeCozyClient,
        googleUtils,
        SOURCE_ACCOUNT_ID
      )
      try {
        await synchronizeContacts(cozyContacts, googleContacts, strategy)
      } catch (err) {
        expect(err).toMatch('Unable to create google contact')
      }
    })

    it('should fail nicely on cozy client error', async () => {
      const googleContacts = []
      fakeCozyClient.save.mockRejectedValue('Unable to save contact in cozy')
      const strategy = getCozyToGoogleStrategy(
        fakeCozyClient,
        googleUtils,
        SOURCE_ACCOUNT_ID
      )
      try {
        await synchronizeContacts(cozyContacts, googleContacts, strategy)
      } catch (err) {
        expect(err).toMatch('Unable to save contact in cozy')
      }
    })
  })
})
