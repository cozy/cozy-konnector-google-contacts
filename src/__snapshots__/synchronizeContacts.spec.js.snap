// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`synchronizeContacts function should synchronize contacts: adanMuellerInCozy 1`] = `
Array [
  Object {
    "_type": "io.cozy.contacts",
    "address": Array [],
    "birthday": null,
    "company": null,
    "cozyMetadata": Object {
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "contactsAccountsId": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
          "id": "people/944070",
          "konnector": "konnector-google",
          "lastSync": "2018-05-05T09:09:00.115Z",
          "remoteRev": "6092c3ca-c9f8-4abb-a27b-3add4158bfc2",
        },
      },
    },
    "email": Array [],
    "metadata": Object {
      "google": Object {
        "metadata": Object {
          "deleted": false,
        },
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Mueller",
      "givenName": "Adan",
    },
    "note": null,
    "phone": Array [],
    "relationships": Object {
      "accounts": Object {
        "data": Array [
          Object {
            "_id": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
            "_type": "io.cozy.contacts.accounts",
          },
        ],
      },
    },
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: clemensRomagueraInCozy 1`] = `
Array [
  Object {
    "_rev": "9a045a8e-8db7-49eb-a0b3-8388566f8a9c",
    "_type": "io.cozy.contacts",
    "address": Array [],
    "birthday": null,
    "company": "Cozy cloud",
    "cozyMetadata": Object {
      "createdAt": "2017-11-20T19:33:00.123Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "contactsAccountsId": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
          "id": "people/56863",
          "konnector": "konnector-google",
          "lastSync": "2018-05-05T09:09:00.115Z",
          "remoteRev": "da455898-cffa-4ba5-988f-606243226509",
        },
      },
      "updatedAt": "2019-02-12T12:18:00.222Z",
      "updatedByApps": Array [
        "Contacts",
      ],
    },
    "email": Array [
      Object {
        "address": "clemens@cozycloud.cc",
        "label": "Personal",
        "primary": true,
        "type": "personal",
      },
    ],
    "id": "clemens-romaguera-edit-conflicts",
    "metadata": Object {
      "google": Object {
        "metadata": Object {
          "deleted": false,
        },
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Romaguera",
      "givenName": "Clemens",
    },
    "note": null,
    "phone": Array [],
    "relationships": Object {
      "accounts": Object {
        "data": Array [
          Object {
            "_id": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
            "_type": "io.cozy.contacts.accounts",
          },
        ],
      },
      "groups": Object {
        "data": Object {
          "_id": "ba36efdf-f09b-41cf-ba6b-eb137214d676",
          "_type": "io.cozy.contacts.groups",
        },
      },
    },
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: colemanHintzInGoogle 1`] = `
Array [
  Object {
    "addresses": Array [],
    "biographies": Array [],
    "birthdays": Array [],
    "emailAddresses": Array [],
    "names": Array [
      Object {
        "familyName": "Hintz",
        "givenName": "Coleman",
      },
    ],
    "organizations": Array [],
    "phoneNumbers": Array [],
  },
  "people/1655899",
  "f673b03e-a077-4a58-8437-465eab2d8283",
]
`;

exports[`synchronizeContacts function should synchronize contacts: destroyAureliaHayesInCozy 1`] = `
Array [
  Object {
    "_rev": "1f0eafe5-fa2b-482c-b9b9-ddb39792ae24",
    "_type": "io.cozy.contacts",
    "cozyMetadata": Object {
      "createdAt": "2017-03-22T07:33:00.123Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "id": "people/672617",
          "remoteRev": "ea79ad60-8c9f-4143-830f-07dfb260630b",
        },
      },
      "updatedAt": "2018-11-12T18:18:00.222Z",
      "updatedByApps": Array [
        "Contacts",
      ],
    },
    "id": "aurelia-hayes-deleted-on-google",
    "name": Object {
      "familyName": "Hayes",
      "givenName": "Aurelia",
    },
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: destroyFabiolaGrozdanaInCozy 1`] = `
Array [
  Object {
    "_rev": "678aad8-fa2b-482c-b9b9-8768bcfe3",
    "_type": "io.cozy.contacts",
    "address": Array [],
    "birthday": null,
    "company": null,
    "cozyMetadata": Object {
      "createdAt": "2017-03-22T07:33:00.123Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "id": "people/924609",
          "remoteRev": "ea79ad60-8c9f-4143-830f-07dfb260630b",
        },
      },
      "updatedAt": "2018-11-12T18:18:00.222Z",
      "updatedByApps": Array [
        "Contacts",
        "konnector-google",
      ],
    },
    "email": Array [],
    "id": "fabiola-grozdana-deleted-on-cozy",
    "metadata": Object {
      "google": Object {
        "metadata": null,
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Grozdana",
      "givenName": "Fabiola",
    },
    "note": null,
    "phone": Array [],
    "trashed": true,
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: destroyJohannaMoenInCozy 1`] = `
Array [
  Object {
    "_rev": "fda8d01b-ff6d-4ce9-b6ea-749a4e3da3e1",
    "_type": "io.cozy.contacts",
    "cozyMetadata": Object {
      "createdAt": "2018-04-22T17:33:00.123Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "id": "people/106666",
          "remoteRev": "bce9d37f-e0f9-44e6-8aa6-51fa8f8781cc",
        },
      },
      "updatedAt": "2018-12-22T15:18:00.222Z",
      "updatedByApps": Array [
        "Contacts",
        "konnector-google",
      ],
    },
    "id": "johanna-moen-deleted-on-google-conflict",
    "name": Object {
      "familyName": "Moen",
      "givenName": "Johanna",
    },
    "trashed": false,
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: destroyRodTrompInCozy 1`] = `
Array [
  Object {
    "_rev": "af1cd883-216b-42eb-9766-d54ebed38658",
    "_type": "io.cozy.contacts",
    "address": Array [],
    "birthday": null,
    "company": null,
    "cozyMetadata": Object {
      "createdAt": "2018-04-22T17:33:00.123Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
      "updatedAt": "2018-12-22T15:18:00.222Z",
      "updatedByApps": Array [
        "Contacts",
        "konnector-google",
      ],
    },
    "email": Array [],
    "id": "deleted-on-cozy-no-resource-name",
    "metadata": Object {
      "google": Object {
        "metadata": null,
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Tromp",
      "givenName": "Rod",
    },
    "note": null,
    "phone": Array [],
    "trashed": true,
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: fabiolaGrozdanaInGoogle 1`] = `
Array [
  "people/924609",
]
`;

exports[`synchronizeContacts function should synchronize contacts: johnDoeInCozy 1`] = `
Array [
  Object {
    "_rev": "9a045a8e-8db7-49eb-a0b3-8388566f8a9c",
    "_type": "io.cozy.contacts",
    "address": Array [],
    "birthday": "1955-2-22",
    "company": "Cozy cloud",
    "cozyMetadata": Object {
      "createdAt": "2016-06-30T09:33:00.123Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "contactsAccountsId": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
          "id": "people/229876",
          "konnector": "konnector-google",
          "lastSync": "2018-05-05T09:09:00.115Z",
          "remoteRev": "6020dd2f-c9b8-4865-bdbd-078faad65204",
        },
      },
      "updatedAt": "2019-01-22T18:18:00.222Z",
      "updatedByApps": Array [
        "Contacts",
      ],
    },
    "email": Array [
      Object {
        "address": "john.doe@posteo.net",
        "primary": false,
        "type": "personal",
      },
      Object {
        "address": "john.doe@cozycloud.cc",
        "primary": true,
      },
    ],
    "id": "john-doe-edited",
    "metadata": Object {
      "google": Object {
        "metadata": null,
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Doe",
      "givenName": "John",
    },
    "note": null,
    "phone": Array [],
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: johnDoeInGoogle 1`] = `
Array [
  Object {
    "addresses": Array [],
    "biographies": Array [],
    "birthdays": Array [
      Object {
        "date": Object {
          "day": "22",
          "month": "2",
          "year": "1955",
        },
        "metadata": Object {
          "primary": true,
        },
      },
    ],
    "emailAddresses": Array [
      Object {
        "metadata": Object {
          "primary": false,
        },
        "type": "personal",
        "value": "john.doe@posteo.net",
      },
      Object {
        "metadata": Object {
          "primary": true,
        },
        "type": undefined,
        "value": "john.doe@cozycloud.cc",
      },
    ],
    "names": Array [
      Object {
        "familyName": "Doe",
        "givenName": "John",
      },
    ],
    "organizations": Array [
      Object {
        "name": "Cozy cloud",
      },
    ],
    "phoneNumbers": Array [],
  },
  "people/229876",
  "9659474d-f3ce-47a5-a9f1-01b55e6b0987",
]
`;

exports[`synchronizeContacts function should synchronize contacts: kayleighYundtInCozy 1`] = `
Array [
  Object {
    "_type": "io.cozy.contacts",
    "address": Array [],
    "birthday": null,
    "company": null,
    "cozyMetadata": Object {
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "contactsAccountsId": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
          "id": "people/751021",
          "konnector": "konnector-google",
          "lastSync": "2018-05-05T09:09:00.115Z",
          "remoteRev": "958e26de-80d2-4051-8ba9-ae954dffd9f6",
        },
      },
    },
    "email": Array [],
    "metadata": Object {
      "google": Object {
        "metadata": Object {
          "deleted": false,
        },
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Yundt",
      "givenName": "Kayleigh",
    },
    "note": null,
    "phone": Array [],
    "relationships": Object {
      "accounts": Object {
        "data": Array [
          Object {
            "_id": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
            "_type": "io.cozy.contacts.accounts",
          },
        ],
      },
    },
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: larueCreminInCozy 1`] = `
Array [
  Object {
    "address": Array [],
    "birthday": null,
    "company": null,
    "cozyMetadata": Object {
      "createdAt": "2014-02-22T09:55:00.222Z",
      "createdByApp": "konnector-google",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "cb31eb3e-989e-4818-8b45-afed904237da",
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "contactsAccountsId": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
          "id": "people/78485",
          "konnector": "konnector-google",
          "lastSync": "2018-05-05T09:09:00.115Z",
          "remoteRev": "6deab8dd-5a0f-451d-bcfe-fd81dd57bdcd",
        },
      },
      "updatedAt": "2018-03-22T12:09:00.222Z",
      "updatedByApps": Array [
        "Contacts",
      ],
    },
    "email": Array [],
    "id": "larue-cremin-attached-to-another-source",
    "metadata": Object {
      "google": Object {
        "metadata": null,
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Cremin",
      "givenName": "Larue",
    },
    "note": null,
    "phone": Array [],
    "relationships": Object {
      "accounts": Object {
        "data": Array [
          Object {
            "_id": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
            "_type": "io.cozy.contacts.accounts",
          },
        ],
      },
    },
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: larueCreminInGoogle 1`] = `
Array [
  Object {
    "addresses": Array [],
    "biographies": Array [],
    "birthdays": Array [],
    "emailAddresses": Array [],
    "names": Array [
      Object {
        "familyName": "Cremin",
        "givenName": "Larue",
      },
    ],
    "organizations": Array [],
    "phoneNumbers": Array [],
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: reinholdJenkinsInCozy 1`] = `
Array [
  Object {
    "address": Array [],
    "birthday": null,
    "company": null,
    "cozyMetadata": Object {
      "createdAt": "2016-05-11T09:09:00.222Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "3.3.0",
      "doctypeVersion": 2,
      "sourceAccount": null,
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "contactsAccountsId": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
          "id": "people/96508",
          "konnector": "konnector-google",
          "lastSync": "2018-05-05T09:09:00.115Z",
          "remoteRev": "963fd240-d568-4b24-9f26-6e658d4e4958",
        },
      },
      "updatedAt": "2018-11-11T09:09:00.222Z",
      "updatedByApps": Array [
        "Contacts",
      ],
    },
    "email": Array [],
    "id": "reinhold-jenkins-no-source",
    "metadata": Object {
      "google": Object {
        "metadata": null,
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Jenkins",
      "givenName": "Reinhold",
    },
    "note": null,
    "phone": Array [],
    "relationships": Object {
      "accounts": Object {
        "data": Array [
          Object {
            "_id": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
            "_type": "io.cozy.contacts.accounts",
          },
        ],
      },
      "groups": Object {
        "data": Object {
          "_id": "ba36efdf-f09b-41cf-ba6b-eb137214d676",
          "_type": "io.cozy.contacts.groups",
        },
      },
    },
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: reinholdJenkinsInGoogle 1`] = `
Array [
  Object {
    "addresses": Array [],
    "biographies": Array [],
    "birthdays": Array [],
    "emailAddresses": Array [],
    "names": Array [
      Object {
        "familyName": "Jenkins",
        "givenName": "Reinhold",
      },
    ],
    "organizations": Array [],
    "phoneNumbers": Array [],
  },
]
`;

exports[`synchronizeContacts function should synchronize contacts: scarlettKundeInCozy 1`] = `
Array [
  Object {
    "_rev": "36bd0782-bda3-4a3b-900a-0d571cc361b6",
    "_type": "io.cozy.contacts",
    "address": Array [],
    "birthday": null,
    "company": null,
    "cozyMetadata": Object {
      "createdAt": "2017-03-22T07:33:00.123Z",
      "createdByApp": "Contacts",
      "createdByAppVersion": "2.0.0",
      "doctypeVersion": 2,
      "sourceAccount": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
      "sync": Object {
        "45c49c15-4b00-48e8-8bfd-29f8177b89ff": Object {
          "contactsAccountsId": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
          "id": "people/364391",
          "konnector": "konnector-google",
          "lastSync": "2018-05-05T09:09:00.115Z",
          "remoteRev": "209006ac-a425-426a-906e-3002d56597fa",
        },
      },
      "updatedAt": "2018-11-12T18:18:00.222Z",
      "updatedByApps": Array [
        "Contacts",
      ],
    },
    "email": Array [],
    "id": "scarlett-gutkowski-edited-on-google",
    "metadata": Object {
      "google": Object {
        "metadata": Object {
          "deleted": false,
        },
      },
      "version": 1,
    },
    "name": Object {
      "familyName": "Kunde",
      "givenName": "Scarlett",
    },
    "note": null,
    "phone": Array [],
    "relationships": Object {
      "accounts": Object {
        "data": Array [
          Object {
            "_id": "45c49c15-4b00-48e8-8bfd-29f8177b89ff",
            "_type": "io.cozy.contacts.accounts",
          },
        ],
      },
    },
  },
]
`;
